### ISSUE TITLE
maybe a bug? different gradients of x**2 and x*x?

### ISSUE THREAD

**Issue by bojone** [tensorflow/tensorflow#20467]
```
import tensorflow as tf

sess = tf.Session()

lr = 0.1

x = tf.Variable(1.)
y = 0.5 * x * x
# y = 0.5 * x**2

P = tf.global_variables()

grads = tf.gradients(y, P)

ops = []
for p,g in zip(P, grads):
    ops.append(tf.assign(p, p - lr * g))

with tf.control_dependencies(ops):
    ops2 = []
    grads = tf.gradients(y, P)
    for p,g in zip(P, grads):
        ops2.append(tf.assign(p, p - lr * g))


sess.run(tf.global_variables_initializer())

sess.run(ops2)
```

if `y = 0.5 * x * x` it return `[0.80499995] ` but `y = 0.5 * x**2` return `[0.81]`, why?

how can I do if I want to run two steps of GD in one sess.run ?

**Comment by bojone** [comment]
there is a similar code.

```
import tensorflow as tf

sess = tf.Session()

lr = 0.1
x = tf.Variable(1.)
y = 0.5 * x * x
# y = 0.5 * x**2

P = tf.global_variables()
grads = tf.gradients(y, P)

ops = []
for p,g in zip(P, grads):
    ops.append(tf.assign(p, p - lr * g))

with tf.control_dependencies(ops):
    p_ = tf.identity(P[0])
    y_ = tf.identity(y)

sess.run(tf.global_variables_initializer())
sess.run([p_, y_])
```
why the final `y_` is 0.5 but `p_` is 0.9 ?  if p is 0.9, should y is `0.5 * 0.9 * 0.9` ?

**Comment by facaiy** [comment]
In your case, I think the execution order of `y` and `assign` operator is non-determined.

```python
with tf.control_dependencies([tf.assign(x, x - 0.1)]):
    y = 0.5 * x * x
    x_ = tf.identity(x)
    y_ = tf.identity(y)

# x_, y_ = [0.9, 0.40499997]
```

**Comment by bojone** [comment]
@facaiy Yet, I can guess the reason. But how can I promise the right order without re-defining the tensor `y` ?

**Comment by bojone** [comment]
somebody has the same confusion about it. like 
https://stackoverflow.com/questions/44352654/tensorflow-gradients-across-post-processing-assign-ops

**Comment by facaiy** [comment]
@bojone Hi, I think it better to ask the question on stackoverflow.

Is the code snippet what you need? If yes, perhaps you will be interested in [ tf.train.Optimizer ](https://www.tensorflow.org/versions/r1.9/api_docs/python/tf/train/Optimizer)

```python
import tensorflow as tf

lr = 0.1

x = tf.get_variable('x',
                    shape=(),
                    initializer=tf.constant_initializer(1.0))
y = 0.5 * x * x
# y = 0.5 * x ** 2

grads = tf.gradients(y, [x])
grad = grads[0]
with tf.control_dependencies([x, grad]):
    update_op = tf.assign(x, x - lr * grad)

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())

    for _ in range(2):
        print('x, y = ', sess.run([x, y]))
        sess.run(update_op)

# x, y =  [1.0, 0.5]
# x, y =  [0.9, 0.40499997]
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by bignamehyp** [comment]
I don't quite understand your confusion.  This is the expected behavior as @facaiy commented.

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this still an issue?
